hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs

```diff
@@ -156,34 +163,49 @@ loadCradle iniDynFlags (NewCradle fp) def action = do
  where
   -- | Initialise the given cradle. This might fail and return an error via `IdeResultFail`.
   -- Reports its progress to the client.
-  initialiseCradle :: (MonadIde m, HasGhcModuleCache m, GHC.GhcMonad m, MonadBaseControl IO m)
-                  => BIOS.Cradle -> (Progress -> IO ()) -> m (IdeResult a)
+  initialiseCradle :: (MonadIde m, HasGhcModuleCache m, GHC.GhcMonad m)
+                  => Bios.Cradle -> (Progress -> IO ()) -> m (IdeResult a)
   initialiseCradle cradle f = do
-    res <- BIOS.initializeFlagsWithCradleWithMessage (Just (toMessager f)) fp cradle
+    res <- Bios.initializeFlagsWithCradleWithMessage (Just (toMessager f)) fp cradle
     case res of
-      BIOS.CradleNone ->
+      Bios.CradleNone ->
         -- Note: The action is not run if we are in the none cradle, we
         -- just pretend the file doesn't exist.
         return $ IdeResultOk def
-      BIOS.CradleFail err -> do
-        logm $ "GhcException on cradle initialisation: " ++ show err
+      Bios.CradleFail (Bios.CradleError code msg) -> do
+        warningm $ "Fail on cradle initialisation: (" ++ show code ++ ")" ++ show msg
+
+        -- Send a detailed diagnostic to the user.
+
+        let normalizedUri = J.toNormalizedUri (filePathToUri fp)
+            sev = Just DsError
+            range = Range (Position 0 0) (Position 1 0)
+            msgTxt =
+              [ "Fail on initialisation for \"" <> Text.pack fp <> "\"."
+              ] <> map Text.pack msg
+            source = Just "bios"
+            diag = Diagnostic range sev Nothing source (Text.unlines msgTxt) Nothing
+
+        liftIO $ publishDiagnostics maxBound normalizedUri Nothing
+            (Map.singleton source (SL.singleton diag))
```

Comment:
This is now the publishing of diagnostics. @alanz I dont understand how we intend to change the `maxToSend` flag.

Reply:
it should not be part of `PublishDiagnostics`.   Rather make the type

```haskell
type PublishDiagnostics = J.NormalizedUri -> J.TextDocumentVersion -> J.DiagnosticsBySource -> IO ()
```

Then the one from `Server.hs` can operate like a callback, where it looks up that value before sending.  Which of course potentially brings in a race condition, but at least it is a read only.

It does leave the way open for the passed in function to be something like `passHandler`

```haskell
passHandler :: TChan ReactorInput -> (a -> FromClientMessage) -> Core.Handler a
passHandler rin c notification = do
  atomically $ writeTChan rin (c notification)
```
which does keep things separate, at the cost of having a different type for the `ReactorInput`.